src/op_utils.rs

```diff
@@ -13,87 +12,269 @@ use num_integer::Integer;
 // We ascribe some additional cost per byte for operations that allocate new atoms
 pub const MALLOC_COST_PER_BYTE: Cost = 10;
 
-pub fn check_arg_count(args: &Node, expected: usize, name: &str) -> Result<(), EvalErr> {
-    if arg_count(args, expected) != expected {
+pub fn get_args<const N: usize>(
+    a: &Allocator,
+    args: NodePtr,
+    name: &str,
+) -> Result<[NodePtr; N], EvalErr> {
+    let mut next = args;
+    let mut counter = 0;
+    let mut ret: [NodePtr; N] = [0; N];
+
+    while let Some((first, rest)) = a.next(next) {
+        next = rest;
+        if counter == N {
+            return err(
+                args,
+                &format!(
+                    "{name} takes exactly {N} argument{}",
+                    if N == 1 { "" } else { "s" }
+                ),
+            );
+        }
+        ret[counter] = first;
+        counter += 1;
+    }
+
+    if counter != N {
         err(
-            args.node,
+            args,
             &format!(
-                "{name} takes exactly {expected} argument{}",
-                if expected == 1 { "" } else { "s" }
+                "{name} takes exactly {N} argument{}",
```

Comment:
No need to change, but the grammar trick I often use is "argument count should be {N}" to avoid plural switching.

Reply:
(In fact, I guess you can't easily change if tests expect this exact message.)

Reply:
yeah, I tried to keep error messages compatible. Although, I don't think there are that many tests to update